### ISSUE TITLE
What is the approximate final lm_loss range for Thinker training?

### ISSUE THREAD

**Issue by candlewill** [MuyeHuang/DuplexOmni#11]
I successfully trained a Thinker model on my own constructed dataset. In TensorBoard, the lm_loss settles around 0.12.
Could you share what the final (or typical) `lm_loss` values were during your training runs? This would help me determine whether the issue is with my data preparation or with modifications I made to the training code.

If you have any loss curve screenshots from TensorBoard (or similar), that would be extremely helpful as well.

Thanks in advance!

<img width="1959" height="684" alt="Image" src="https://github.com/user-attachments/assets/83c9fc11-11de-4179-9a6b-e4c1d7f628f9" />

**Comment by MuyeHuang** [comment]
> I successfully trained a Thinker model on my own constructed dataset. In TensorBoard, the lm_loss settles around 0.12. Could you share what the final (or typical) `lm_loss` values were during your training runs? This would help me determine whether the issue is with my data preparation or with modifications I made to the training code.
> 
> If you have any loss curve screenshots from TensorBoard (or similar), that would be extremely helpful as well.
> 
> Thanks in advance!
> 
> <img alt="Image" width="1959" height="684" src="https://private-user-images.githubusercontent.com/4916563/636018962-83c9fc11-11de-4179-9a6b-e4c1d7f628f9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODY5Mzc0OTksIm5iZiI6MTc4NjkzNzE5OSwicGF0aCI6Ii80OTE2NTYzLzYzNjAxODk2Mi04M2M5ZmMxMS0xMWRlLTQxNzktOWE2Yi1lNGMxZDdmNjI4ZjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDgxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA4MTdUMDMyNjM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDZkMmRhOGMwNTVjM2MxNzkxNGE4ZWE2YzJkZmQwNzQ0NmNiMTYwM2JiZmIwMGUwOWNhODBiOWNiNGRkMjVmZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.itBzMN2RwCHc-8qewaF33dnB141N7Cm4D9hNuCOM3Pc">

Congratulations on successfully completing the Thinker fine-tuning on your own dataset!

If you are fine-tuning only the Thinker, a final `lm_loss` of around 0.02～0.1 is within the normal range in our experience. Therefore, the final value of approximately 0.12 itself is not concerning.

**However, after looking at your loss curve, the initial loss above 1.0 appears highly unusual. In our experience with real human-interaction data—even though its distribution differs substantially from the fully synthetic training data—the initial `lm_loss` is typically only around 0.15.**

This may indicate that the temporal sequence was incorrectly aligned, or that the dataset was not constructed according to the original data conventions. I would recommend carefully checking the alignment between the streamed inputs and training targets, as well as the target labels and loss masks. A final loss of 0.12 alone does not necessarily confirm that the data pipeline is correct, since the model may eventually fit incorrectly constructed targets.

If you are concerned about your modifications to the training code, you can also fine-tune the Thinker using the ms-swift implementation as a sanity-check baseline. If the same dataset still starts with a loss above 1.0 in ms-swift, the issue is more likely in the data construction; otherwise, it is more likely related to your code modifications.
@candlewill

**Comment by candlewill** [comment]
Appreciate the insights! I’ll go back and carefully review the data construction and training process.
